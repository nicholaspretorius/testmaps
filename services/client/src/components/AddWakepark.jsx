import React, { Component } from "react";
import { Redirect } from "react-router-dom";
import { Formik } from "formik";
import * as Yup from "yup";

import localStorage from "../services/localStorage";

class AddWakepark extends Component {
  state = {};

  onHandleAddWakepark(values) {
    console.log("Values: ", values);
  }

  render() {
    if (!localStorage.isPermitted("post:cableparks")) {
      return <Redirect to="/" />;
    }

    return (
      <div>
        <h3 className="title">Add Wakepark</h3>
        <hr />
        <br />
        <Formik
          initialValues={{
            name: "",
            lat: "",
            lng: "",
            instagramHandle: ""
          }}
          onSubmit={(values, { setSubmitting, resetForm }) => {
            this.onHandleAddWakepark(values);
            resetForm();
            setSubmitting(false);
          }}
          validationSchema={Yup.object().shape({
            name: Yup.string().required("Please enter a name"),
            lat: Yup.string().required("Please enter a latitude"),
            lng: Yup.string().required("Please enter a longitude"),
            instagramHandle: Yup.string()
          })}
        >
          {props => {
            const {
              values,
              touched,
              errors,
              isSubmitting,
              handleChange,
              handleBlur,
              handleSubmit
            } = props;
            return (
              <form onSubmit={handleSubmit}>
                <div className="field">
                  <label htmlFor="input-name" className="label">
                    Name
                  </label>
                  <input
                    type="text"
                    name="name"
                    id="input-name"
                    className={errors.name && touched.name ? "input error" : "input"}
                    placeholder="Enter the wakepark name"
                    value={values.name}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                  {errors.name && touched.name && (
                    <div className="input-feedback" data-testid="errors-name">
                      {errors.name}
                    </div>
                  )}
                </div>
                <div className="field">
                  <label htmlFor="input-lat" className="label">
                    Latitude
                  </label>
                  <input
                    type="text"
                    name="lat"
                    id="input-lat"
                    className={errors.lat && touched.lat ? "input error" : "input"}
                    placeholder="Enter the wakepark latitude location"
                    value={values.lat}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                  {errors.lat && touched.lat && (
                    <div className="input-feedback" data-testid="errors-lat">
                      {errors.lat}
                    </div>
                  )}
                </div>
                <div className="field">
                  <label htmlFor="input-lng" className="label">
                    Longitude
                  </label>
                  <input
                    type="text"
                    name="lng"
                    id="input-lng"
                    className={errors.lng && touched.lng ? "input error" : "input"}
                    placeholder="Enter the wakepark longitude location"
                    value={values.lng}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                  {errors.lng && touched.lng && (
                    <div className="input-feedback" data-testid="errors-lng">
                      {errors.lng}
                    </div>
                  )}
                </div>

                <div className="field">
                  <label htmlFor="input-instagram-handle" className="label">
                    Instagram Handle
                  </label>
                  <input
                    type="text"
                    name="instagram-handle"
                    id="input-instagram-handle"
                    className={
                      errors.instagramHandle && touched.instagramHandle ? "input error" : "input"
                    }
                    placeholder="Enter the wakepark Instagram handle"
                    value={values.instagramHandle}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                  {errors.instagramHandle && touched.instagramHandle && (
                    <div className="input-feedback" data-testid="errors-instagram-handle">
                      {errors.instagramHandle}
                    </div>
                  )}
                </div>

                <input
                  type="submit"
                  className="button is-primary"
                  value="Submit"
                  disabled={isSubmitting}
                />
              </form>
            );
          }}
        </Formik>
      </div>
    );
  }
}

export default AddWakepark;
