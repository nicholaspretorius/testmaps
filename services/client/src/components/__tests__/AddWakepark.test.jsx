import React from "react";
import { cleanup, fireEvent } from "@testing-library/react";

import AddWakepark from "./../AddWakepark.jsx";

afterEach(cleanup);

describe("renders", () => {
  it("properly", () => {
    const { getByText } = renderWithRouter(<AddWakepark />);
    expect(getByText("Add Wakepark")).toHaveClass("title");
  });

  it("without values", () => {
    const { getByLabelText, getByText } = renderWithRouter(<AddWakepark />);

    const nameInput = getByLabelText("Name");
    expect(nameInput).toHaveAttribute("type", "text");
    expect(nameInput).not.toHaveValue();

    const latInput = getByLabelText("Latitude");
    expect(latInput).toHaveAttribute("type", "text");
    expect(latInput).not.toHaveValue();

    const lngInput = getByLabelText("Longitude");
    expect(lngInput).toHaveAttribute("type", "text");
    expect(lngInput).not.toHaveValue();

    const instaHandleInput = getByLabelText("Instagram Handle");
    expect(instaHandleInput).toHaveAttribute("type", "text");
    expect(instaHandleInput).not.toHaveValue();

    const buttonInput = getByText("Submit");
    expect(buttonInput).toHaveClass("button is-primary");
    expect(buttonInput).toHaveValue("Submit");
  });

  it("a snapshot", () => {
    const { asFragment } = renderWithRouter(<AddWakepark />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("handles form validation correctly", () => {
  it("when fields are empty", async () => {
    const { getByLabelText, container, findByTestId } = renderWithRouter(<AddWakepark />);

    const form = container.querySelector("form");
    const nameInput = getByLabelText("Name");
    const latInput = getByLabelText("Latitude");
    const lngInput = getByLabelText("Longitude");
    const instaHandleInput = getByLabelText("Instagram Handle");

    fireEvent.blur(nameInput);
    fireEvent.blur(latInput);
    fireEvent.blur(lngInput);
    fireEvent.blur(instaHandleInput);

    expect((await findByTestId("errors-name")).innerHTML).toBe("Please enter a name");
    expect((await findByTestId("errors-lat")).innerHTML).toBe("Please enter a latitude");
    expect((await findByTestId("errors-lng")).innerHTML).toBe("Please enter a longitude");

    fireEvent.submit(form);
  });
});
