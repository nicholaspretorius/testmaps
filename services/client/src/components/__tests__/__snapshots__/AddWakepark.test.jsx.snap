// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3
      class="title"
    >
      Add Wakepark
    </h3>
    <hr />
    <br />
    <form>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-name"
        >
          Name
        </label>
        <input
          class="input"
          id="input-name"
          name="name"
          placeholder="Enter the wakepark name"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-lat"
        >
          Latitude
        </label>
        <input
          class="input"
          id="input-lat"
          name="lat"
          placeholder="Enter the wakepark latitude location"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-lng"
        >
          Longitude
        </label>
        <input
          class="input"
          id="input-lng"
          name="lng"
          placeholder="Enter the wakepark longitude location"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-instagram-handle"
        >
          Instagram Handle
        </label>
        <input
          class="input"
          id="input-instagram-handle"
          name="instagram-handle"
          placeholder="Enter the wakepark Instagram handle"
          type="text"
          value=""
        />
      </div>
      <input
        class="button is-primary"
        type="submit"
        value="Submit"
      />
    </form>
  </div>
</DocumentFragment>
`;
